File: solutions/go/bottle-song/1/bottle_song.go
package bottlesong

import (
	"fmt"
	"unicode"
)

type verse string

var numbersMapping = map[int]string{
	0:  "no",
	1:  "one",
	2:  "two",
	3:  "three",
	4:  "four",
	5:  "five",
	6:  "six",
	7:  "seven",
	8:  "eight",
	9:  "nine",
	10: "ten",
}

func Recite(startBottles, takeDown int) (verses []string) {
	stopBottles := startBottles - takeDown
	for i := startBottles; i > stopBottles; i-- {
		capitalized := fmt.Sprintf(
			"%s %s hanging on the wall,",
			capitalize(numbersMapping[i]),
			greenBottle(i),
		)

		verses = append(verses, capitalized)
		verses = append(verses, capitalized)
		verses = append(verses, "And if one green bottle should accidentally fall,")
		verses = append(verses, fmt.Sprintf("There'll be %s %s hanging on the wall.", numbersMapping[i-1], greenBottle(i-1)))
		if i-1 > stopBottles {
			verses = append(verses, "")
		}
	}

	return
}

func greenBottle(n int) string {
	if n == 1 {
		return "green bottle"
	}
	return "green bottles"
}

func capitalize(s string) string {
	r := []rune(s)
	r[0] = unicode.ToUpper(r[0])
	return string(r)
}

File: solutions/elixir/bottle-song/1/lib/bottle_song.ex
defmodule BottleSong do
  @moduledoc """
  Handles lyrics of the popular children song: Ten Green Bottles
  """

  @numbers_mapping 0..10
    |> Enum.zip(["no", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"])
    |> Enum.into(%{}, fn {k, v} -> {k, v} end)

  @spec recite(pos_integer, pos_integer) :: String.t()
  def recite(start_bottle, take_down) do
    start_bottle..(start_bottle - take_down + 1)//-1
    |> Enum.reduce([], fn count, acc -> acc ++ [lyrics(count)] end)
    |> Enum.join("\n\n")
  end

  defp lyrics(count) do
    nbr_str = String.capitalize(@numbers_mapping[count])
    "#{nbr_str} #{green_bottle(count)} hanging on the wall,\n" <>
    "#{nbr_str} #{green_bottle(count)} hanging on the wall,\n" <>
    "And if one #{green_bottle(1)} should accidentally fall,\n" <>
    "There'll be #{@numbers_mapping[count - 1]} #{green_bottle(count - 1)} hanging on the wall."
  end

  defp green_bottle(1), do: "green bottle"
  defp green_bottle(_), do: "green bottles"
end
